python/lsst/afw/detection/multiband.py

```diff
@@ -0,0 +1,246 @@
+# This file is part of afw.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+__all__ = ["MultibandFootprint"]
+
+import numpy as np
+
+from lsst.geom import Point2I
+from lsst.afw.geom import SpanSet
+from lsst.afw.image import Mask, Image, MaskedImage, MultibandImage
+from lsst.afw.multiband import MultibandBase
+from . import Footprint, makeHeavyFootprint
+
+
+class MultibandFootprint(MultibandBase):
+    """Multiband Footprint class
+
+    A `MultibandFootprint` is a collection of HeavyFootprints that have
+    the same `SpanSet` and `peakCatalog` but different flux in each band.
+    """
+    def __init__(self, filters, singles=None, images=None, footprint=None, xy0=None, peaks=None, thresh=0):
+        """Initialize a `MultibandFootprint` object
+
+        Parameters
+        ----------
+        filters: list
+            List of filter names.
+            If `singles` is an `OrderedDict` then this argument is ignored,
+            otherwise it is required.
+        singles: list
+            A list of single band `HeavyFootprint` objects.
+            Either `singles` or `images` must be specified.
+            If `singles` is not `None`, then all arguments other than
+            `filters` are ignored.
+            Each `HeavyFootprint` should have the same `PeakCatalog` but
+            is allowed to have a different `SpanSet`, in which case the
+            `SpanSet` for each band is combined into the `SpanSet` of the
+            `MultibandFootprint`.
```

Comment:
Unlike `MultibandImage`, where each single band instance has the same shape and bounding box, each footprint in `MultibandFootprints` can have an array that is a different size, so it cannot be built into a higher dimensional array directly. The full 3D array is only created for `MultibandFootprints` if the `getImage` or `getArray` method is called.

We could build a 3D array using the footprint that is the union of all of the single band footprints (so the array in each band has the same length), but then we have to find a quick and clever method pad each 1D `HeavyFootprint` `array` with zeros in the appropriate places.

Another option is to always create and store the full  padded 3D multiband array (the result of `getArray`) and slice it appropriately to create each `single`, but I wonder if that is adding unnecessary overhead.

But since we can't really slice the footprints, I think the simplest (and fastest?) method is to just store the single band `HeavyFootprints` and only build the higher dimensional images if asked (ie. the current implementation). Do you agree?

Reply:
Permitting `MultibandFootprint` to have a single higher-dimensional array or multiple differently-sized 1-d arrays makes sense (having different `Footprint`s in bands makes me question *a bit* whether those belong in the same container, but this should be driven more by use cases than my sense of cleanliness).

However, could we make it so a `MultibandFootprint` is always one or the other after construction, rather than permitting the array to be allocated on request?  Operations that modify an object while seeming to just be getters are often problematic.

Reply:
> having different Footprints in bands makes me question a bit whether those belong in the same container

I think that I agree with you on this. I've gone back and forth a lot about whether or not the footprints should all have the same `SpanSet`. Peter and I had a similar conundrum when we were building the difference kernels: do we truncate all of them to have the same footprint or cut them off at the same threshold with different footprints? In that case we built the difference kernels using scarlet while enforcing L0 sparsity and surprisingly, the best fit to the data uses PSF footprints with a different size in each band. The result is that while the model in each scarlet band is based on a morphology that will have the same footprint, convolving the morphology with the PSF in each band results in footprints that have different sizes.

But there doesn't seem to be any harm in just using the union of all the resulting scarlet footprints and padding them with zeros when we create the `HeavyFootprints`, which is my recommended solution (and is also the way I construct a `MultibandFootprint` from a collection of thresholded images), but I felt like we should support different sizes. Your hesitation coupled with my own qualms about this leads me to believe that maybe we should require all of the single band footprints to have the same `SpanSet` and force the user to decide how to best come up with matched footprints (ie. pad them with zeros or use the union of all the `SpanSets` to extract a `HeavyFootprint` with the same `SpanSet` from a collection of multiband images). In the case of scarlet both methods give the same result (since the models don't have any noise) but for more general footprints there can be a difference.

Using the same footprint makes it easy to build and store a higher dimensional array and `MultibandFootprint` will be more similar in structure to the other multiband classes, so if you agree then I'll update the code to require the input footprints to have the same `SpanSets`.

Reply:
Sounds reasonable to me.  People can obviously still have different single-band HeavyFootprints in different bands, and just not use MultibandFootprint, and I think that's okay, as they wouldn't have gained as much from doing so.